aptos-move/aptos-e2e-comparison-testing/src/data_collection.rs

```diff
@@ -0,0 +1,262 @@
+        };
+
+        if !is_aptos_package(&package_name)
+            && !compilation_cache
+                .compiled_package_map
+                .contains_key(&package_info)
+        {
+            if compilation_cache.failed_packages.contains(&package_info) {
+                return None;
+            }
+            let res = dump_and_compile_from_package_metadata(
+                package_info.clone(),
+                current_dir,
+                &map,
+                compilation_cache,
+                None,
+            );
+            if res.is_err() {
+                println!("compile package failed at:{}", version);
+                return None;
+            }
+        }
+        Some(package_info)
+    }
+
+    fn dump_txn_index(
+        data_manager: &mut DataManager,
+        txn_index: TxnIndex,
+        data_state: &HashMap<StateKey, StateValue>,
+        epoch_result_res: Result<Vec<TransactionOutput>>,
+        dump_write_set: bool,
+    ) {
+        // dump TxnIndex
+        data_manager.dump_txn_index(txn_index.version, &txn_index);
+        // dump state data
+        data_manager.dump_state_data(txn_index.version, data_state);
+        // dump write set
+        if dump_write_set {
+            let output = epoch_result_res.unwrap();
+            assert_eq!(output.len(), 1);
+            let write_set = output[0].write_set();
+            data_manager.dump_write_set(txn_index.version, write_set);
+        }
+    }
+
+    pub async fn dump_data(&self, begin: Version, limit: u64) -> Result<()> {
+        println!("begin dumping data");
+        let compilation_cache = Arc::new(Mutex::new(CompilationCache::default()));
+        let data_manager = Arc::new(Mutex::new(DataManager::new_with_dir_creation(
+            &self.current_dir,
+        )));
+        let index_writer = Arc::new(Mutex::new(IndexWriter::new(&self.current_dir)));
+
+        let mut cur_version = begin;
+
+        while cur_version < begin + limit {
+            let batch = if cur_version + self.batch_size <= begin + limit {
+                self.batch_size
+            } else {
+                begin + limit - cur_version
+            };
+            let res_txns = self
+                .debugger
+                .get_and_filter_committed_transactions(cur_version, batch, self.filter_condition)
+                .await;
+            // if error happens when collecting txns, log the version range
+            if res_txns.is_err() {
+                index_writer
+                    .lock()
+                    .unwrap()
+                    .write_err(&format!("{}:{}", cur_version, batch));
+            }
+            let txns = res_txns.unwrap_or_default();
+            if !txns.is_empty() {
+                let mut txn_execution_ths = vec![];
+                for (version, txn, source_code_data) in txns {
+                    println!("get txn at version:{}", version);
+
+                    let compilation_cache = compilation_cache.clone();
```

Comment:
Are the clonings here needed? Or rather, why do we need to move those values around?

Reply:
It is of Arc type so I have to clone the pointer so that it can be shared among threads. Is there a better way to do it?